sound/soc/sof/ipc4-topology.c

```diff
@@ -1222,6 +1222,69 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
 }
 #endif
 
+static int ipc4_set_fmt_mask(struct snd_mask *fmt, unsigned int bit_depth)
+{
+	switch (bit_depth) {
+	case 16:
+		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S16_LE);
+		break;
+	case 24:
+		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S24_LE);
+		break;
+	case 32:
+		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int ipc4_copier_set_capture_fmt(struct snd_sof_dev *sdev,
+				       struct snd_pcm_hw_params *pipeline_params,
+				       struct snd_pcm_hw_params *fe_params,
+				       struct sof_ipc4_available_audio_format *available_fmt)
+{
+	struct sof_ipc4_audio_format *audio_fmt;
+	unsigned int sample_valid_bits;
+	bool multiple_formats = false;
+	bool fe_format_match = false;
+	struct snd_mask *fmt;
+	int i;
+
+	for (i = 0; i < available_fmt->num_output_formats; i++) {
+		unsigned int val;
+
+		audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
+		val = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);
+
+		if (i == 0)
+			sample_valid_bits = val;
+		else if (sample_valid_bits != val)
+			multiple_formats = true;
+
+		if (snd_pcm_format_width(params_format(fe_params)) == val)
+			fe_format_match = true;
+	}
```

Comment:
I think if I look at this loop and the one in `sof_ipc4_pcm_dai_link_fixup_rate()` they are doing exactly the same thing, one with rate the other with valid_bits, yet the implementation of them are unique.
I would consider using  the same logical flow in both to make things simpler.

Reply:
Right, the verbs are negated in between the two, but they do the same, if I'm not mistaken?

Reply:
Yes, I got that part, what I'm saying that the two loops can be made identical looking, in essence the only diff is _what_ is compared, but the flow can be the same

```c
	for (i = 0; i < available_fmt->num_output_formats; i++) {
		unsigned int val;

		audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
		val = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);

		if (i == 0)
			sample_valid_bits = val;
		else if (sample_valid_bits != val)
			multiple_formats = true;

		if (snd_pcm_format_width(params_format(fe_params)) == val)
			fe_format_match = true;
	}
```
VS (this is what you have in `sof_ipc4_pcm_dai_link_fixup_rate()`)
```c
	audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
	sample_valid_bits = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);

	for (i = 0; i < available_fmt->num_output_formats; i++) {
		unsigned int val;

		audio_fmt = &available_fmt->output_pin_fmts[i].audio_fmt;
		val = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(audio_fmt->fmt_cfg);

		if (snd_pcm_format_width(params_format(fe_params)) == val)
			fe_format_match = true;

		if (sample_valid_bits != val)
			single_formats = false;
	}
```

right, this is nitpicking.

Reply:
@ujfalusi please check, tried to use same style in both functions now.

Reply:
Thanks, looks confusingly similar :D